### ISSUE TITLE
Support filtering connections from slirp4netns 

### ISSUE THREAD

**Issue by KCSesh** [containers/podman#10297]
/kind feature

**Description**

Is there any support for filtering connections from slirp4netns? 
https://github.com/rootless-containers/slirp4netns/blob/master/slirp4netns.1.md#filtering-connections

I would like to [allow_host_loopback](http://docs.podman.io/en/latest/markdown/podman-run.1.html#network-mode-net)
But then setup additional iptable rules to **not** allow for total host access. i.e. allow loopback to host on a specific port and set this on a per container basis. 

Is there any way to do this today? 
I imagine a custom prestart hook could likely be written to enter the network namespace and setup iptable rules. 
The only problem is knowing the network namespace path at hook execution time.

**Comment by mheon** [comment]
I don't believe we can do this in the container, and Rootless Podman does not have permission to modify the system firewall, so you may have to make these changes manually. @AkihiroSuda @giuseppe Can you confirm?

**Comment by KCSesh** [comment]
Is there a way I can manually make these changes @mheon from the host using root?

**Comment by Luap99** [comment]
This should work to join the container network namespace:
```
$ podman run --name test -d alpine top
# Note this only works when the container is running or you used podman container init test before
$ nspath=$(podman container inspect --format "{{.NetworkSettings.SandboxKey}}" test)
# This command executes your shell in the user and network namespace
$ podman unshare nsenter --net=$nspath
# Here should be able to use iptables
$ iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
```

Note: If you run into `Fatal: can't open lock file /run/xtables.lock: Permission denied` set the  `XTABLES_LOCKFILE` envar to a user writeable location, e.g. `XTABLES_LOCKFILE=/tmp/xtables.lock`.

Also the rules won't persist if the container is stopped and started again.

**Comment by basilgello** [comment]
I am using the `createContainer` OCI hook to add iptables rules.

On simple containers like Debian it works like a charm, but on some containers Podman manages to lauch slirp4netns right before startContainer, but not before createContainer hook is called.

I guess we must ensure that createContainer hook is called after slirp4netns is started.

Also the SandboxKey from the alpine example is non-empty only if no `--userns=keep-id` specified.

**Comment by rhatdan** [comment]
@mheon @Luap99 what is going on with this one?

**Comment by mheon** [comment]
As described, this sounds like an issue with containers with user namespaces, which configure their networking in a different order than other containers (`PostConfigureNetNS` internally). However, to the best of my knowledge, this causes containers to create the network stack *later* than normal, which does not match what @basilgello is seeing.

**Comment by mheon** [comment]
As for a fix, I don't really know what we can do. We can't configure the netns any earlier for containers in a user namespace as we need to create the container in the runtime first, to create the user namespace itself, as the network namespace must be a child of the userns.

**Comment by basilgello** [comment]
> I don't really know what we can do

I think we can rename that setting to reflect it is relevant for rootless mode only and document that. Something like podman-specific rootless-only option. And guard that callback for rootless only.

**Comment by rhatdan** [comment]
@mheon @Luap99 or @basilgello would you like to open a PR to make this happen?

**Comment by basilgello** [comment]
I recall it was pretty easy to implement @rhatdan At least I did some research drafting #10704 I can try finding my notes on the subject this week.

**Comment by rhatdan** [comment]
The PR has stalled, anyone interested in picking this up?

**Comment by rhatdan** [comment]
@AkihiroSuda WDYT?

**Comment by rhatdan** [comment]
Interested in driving it to completion?

**Comment by basilgello** [comment]
@rhatdan @AkihiroSuda

> #10704 looks good

I stalled that PR because the OCI hook is not called as expected in all scenarios, so merging it can lead to a confusion and false sense of security. Since my goal was exactly filtering in rootless mode (i.e slirp4netns) I think moving the PR on slirp4netns side is logically sound.

I will happily test the slirp4netns PR but I am not sure if I get enough time to code it with all the tests.
